### DISCUSSION TITLE
[Idea] Provide a simply product presenter for modules developers

### DISCUSSION THREAD

**Discussion by didpoule** [PrestaShop/PrestaShop#32268] [Ideas]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing features request](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3AImprovement%2CFeature%2CDeveloper+Feature+) and found no previous suggestion of this feature.

### Is your feature request related to a problem?

Using product presenter is pretty verbose and not optimized for now.
Its always complicated for devs to use it because we need to instantiate a factory, a product assembler, get the presentation setting to be able presenting a product.

That's source of error and not pretty to repeat the same  pattern many times.

### Describe the solution you'd like

Simplify the code necessary to present products in modules in this way:
 Wrap the ProductListingPresenter instantiation in a class which manage dependencies itself

### Alternatives you've considered

no

### Additional context

Here an exemple of current usage of ProductPresenter in a module
```php
        $sql = 'SELECT * FROM `'._DB_PREFIX_.'products`';
        $products = Db::getInstance()->executeS($sql);

        if ($products) {
            $assembler = new ProductAssembler($this->context);
            $presenterFactory = new ProductPresenterFactory($this->context);
            $presentationSettings = $presenterFactory->getPresentationSettings();
            $presenter = new ProductListingPresenter(
                new ImageRetriever(
                    $this->context->link
                ),
                $this->context->link,
                new PriceFormatter(),
                new ProductColorsRetriever(),
                $this->context->getTranslator()
            );

            $presentedProducts = [];

            foreach ($products as $key => $rawProduct) {
                $presentedProducts[] = $presenter->present(
                    $presentationSettings,
                    $assembler->assembleProduct($rawProduct),
                    $this->context->language
                );
            }
        }
```

After reworking it might look like this
```php
    $collection = new PrestaShopCollection(Product::class, $this->context->language->id);
        $collection->where('active', '=', 1);
        $products = $collection->getResults();
        $presentedProducts = \PrestaShop\PrestaShop\Adapter\Presenter\Product\ProductSimplePresenter::getInstance($this->context)->presentProducts($products);
```

### Do you plan to work on this feature?

- [X] I'm willing to contribute a [formal specification](https://github.com/PrestaShop/prestashop-specs).
- [X] I'm willing to provide any wireframes or design assets required for this feature.
- [X] I'm willing to submit a Pull Request that implements this feature.
- [X] I'm willing to help verify that the implemented feature works as intended and produces no unintended side effects.

**Comment by didpoule** [comment]
Here a sample module to test the feature
[samplemodule.zip](https://github.com/PrestaShop/PrestaShop/files/9210780/samplemodule.zip)

**Comment by HanaRebaiQA** [comment]
Hello @didpoule ,

Thank you for your suggestion and for your PR. The Maintainer Team will take it into consideration for future developments. 

Please be aware that there is no guarantee that this feature will be developed anytime soon. If this is important to you, we strongly suggest you consider [contacting a professional](https://www.prestashop-project.org/support/) to help you create it. 

If you or anyone else is willing to [contribute this feature](https://devdocs.prestashop.com/1.7/contribute/contribute-pull-requests/) to the PrestaShop Project, the teams will be happy to help by ensuring that it fits well within the project. Please let us know if that is the case!

Thank You!

**Comment by PululuK** [comment]
Hi @HanaRebaiQA 
I think it is a dev feature, so I update the labels.

ping @PrestaShop/prestashop-maintainers 

Good ideia @didpoule, thanks !

**Comment by HanaRebaiQA** [comment]
Thank You @PululuK

**Comment by matks** [comment]
Hello @didpoule well ... I am sorry but I am not 100% convinced by the idea 😅 

You say
```
        $sql = 'SELECT * FROM `'._DB_PREFIX_.'products`';
        $products = Db::getInstance()->executeS($sql);

        if ($products) {
            $assembler = new ProductAssembler($this->context);
            $presenterFactory = new ProductPresenterFactory($this->context);
            $presentationSettings = $presenterFactory->getPresentationSettings();
            $presenter = new ProductListingPresenter(
                new ImageRetriever(
                    $this->context->link
                ),
                $this->context->link,
                new PriceFormatter(),
                new ProductColorsRetriever(),
                $this->context->getTranslator()
            );

            $presentedProducts = [];

            foreach ($products as $key => $rawProduct) {
                $presentedProducts[] = $presenter->present(
                    $presentationSettings,
                    $assembler->assembleProduct($rawProduct),
                    $this->context->language
                );
            }
        }
```
that this is too much.

So if I read PR https://github.com/PrestaShop/PrestaShop/pull/29194/ you just put all of this code inside a class and use it 😅 . Is it really efficient? Instead of writing 30 lines in your module, you write 30 lines in PR https://github.com/PrestaShop/PrestaShop/pull/29194/ and you add them in the Core 😅 I am not sure of the benefit for everyone here

**Comment by didpoule** [comment]
Hi @matks, thanks for your feedback.
In a web agency context, if you need to present products in several places, you have to repeat the whole verbose code I shared. The purpose is just wrapping that code into a singleton class to be able to present products in a single code line instead of copy/paste the initialization.
Furthermore, I added some static cache to prevent multiple presenting for a same product in an exec.

Let me know if you have any question.

**Comment by PululuK** [comment]
> So if I read PR https://github.com/PrestaShop/PrestaShop/pull/29194 you just put all of this code inside a class and use it sweat_smile . Is it really efficient? Instead of writing 30 lines in your module, you write 30 lines in PR https://github.com/PrestaShop/PrestaShop/pull/29194 and you add them in the Core sweat_smile I am not sure of the benefit for everyone here

Hi @matks consider the case where you have several modules.

**Comment by kpodemski** [comment]
It reminds me a bit of a laravel-way of thinking where we want to help developers produce more, quicker, with less code. The truth is that using this code is very common, and most of the time, there's no need to modify the context for `ProductAssembler`, etc. Actually, I don't remember the last time I had to change this part of the code 🤔
